aten/src/ATen/native/cpu/PowKernel.cpp

```diff
@@ -34,126 +36,98 @@ void pow_tensor_tensor_kernel(TensorIteratorBase& iter) {
-          [](scalar_t base) -> scalar_t {
-            return base * base;
-          },
-          [](Vec base) -> Vec { return base * base; }
-        );
-      } else if (exp == 3) {
-        cpu_kernel_vec(iter,
-          [](scalar_t base) -> scalar_t {
-            return base * base * base;
-          },
-          [](Vec base) -> Vec { return base * base * base; }
-        );
-      } else if (exp == -0.5) {
-        cpu_kernel_vec(iter,
-          [](scalar_t base) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
-            return 1.0 / std::sqrt(base);
-          },
-          [](Vec base) -> Vec { return base.rsqrt(); }
-        );
-      } else if (exp == -1) {
-        cpu_kernel_vec(iter,
-          [](scalar_t base) -> scalar_t {
-            return 1.0 / base;
-          },
-          [](Vec base) -> Vec { return base.reciprocal(); }
-        );
-      } else if (exp == -2) {
-        cpu_kernel_vec(iter,
-          [](scalar_t base) -> scalar_t {
-            return 1.0 / (base * base);
-          },
-          [](Vec base) -> Vec { return (base * base).reciprocal(); }
-        );
-      } else {
-        cpu_kernel_vec(iter,
-          [=](scalar_t base) -> scalar_t {
-            return std::pow(base, exp);
-          },
-          [=](Vec base) -> Vec { return base.pow(exp); }
-        );
-      }
+// Floating and complex types (minus Half and BFloat16) allow for optimized kernels
+template <typename scalar_t, typename exp_scalar_t>
+void pow_tensor_scalar_optimized_kernel(TensorIteratorBase& iter, const exp_scalar_t exp) {
+  using Vec = Vec256<scalar_t>;
+  if (exp == 0.5) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) -> scalar_t { return std::sqrt(base); },
+        [](Vec base) -> Vec { return base.sqrt(); });
+  } else if (exp == 2.0) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) -> scalar_t { return base * base; },
+        [](Vec base) -> Vec { return base * base; });
+  } else if (exp == 3.0) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) -> scalar_t { return base * base * base; },
+        [](Vec base) -> Vec { return base * base * base; });
+  } else if (exp == -0.5) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
+          return scalar_t(1.0) / std::sqrt(base);
+        },
+        [](Vec base) -> Vec { return base.rsqrt(); });
+  } else if (exp == -1.0) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) -> scalar_t { return scalar_t(1.0) / base; },
+        [](Vec base) -> Vec { return base.reciprocal(); });
+  } else if (exp == -2.0) {
+    cpu_kernel_vec(
+        iter,
+        [](scalar_t base) -> scalar_t { return scalar_t(1.0) / (base * base); },
+        [](Vec base) -> Vec { return (base * base).reciprocal(); });
+  } else {
+    cpu_kernel_vec(
+        iter,
+        [=](scalar_t base) -> scalar_t {
+          return std::pow(base, scalar_t(exp));
```

Comment:
In the current code, even a `float` tensor uses `double` exp for better precision.
I think this change has been causing some `pow_test` failures.
I'll check if `scalar_t(exp)` can be changed to `exp`.

Reply:
If I remove `complex support` from `pow`  in this PR, and replace `scalar_t(exp)` with `exp` in line numbers 79 & 82, then `pow_test` passes.

This means `pow_test` tests are failing due to decreasing the precision when `iter.common_dtype()` is `float` and exponent is `double`, as `scalar_t(exp)` changes the double exponent to float.

Reply:
3 tests in `pow_test` are failing due to decreasing the precision when `iter.common_dtype()` is `float` and exponent is `double`, as `scalar_t(exp)` changes the double exponent to float, which is undesired. Is there any way to avoid this?
One possible solution is to create two separate kernels for the else case, but that would make the code look weird.

Reply:
I was worried this problem would occur but wanted to let the tests run first haha. Let me see how we could handle this. Thank you for investigating the bug!

Reply:
The only solution I have so far is adding another template type cast_scalar_t. I tried using `c10::is_complex` to create diverging paths but the compiler is complaining. @mruberry is there a better way to handle this?

Reply:
Ah, I missed the templatization idea again. Adding the template type `cast_scalar_t` only requires modifying 5 lines.
A comment can be added to explain why another template type being added.

Reply:
@imaginary-person if `BFloat16` can benefit from the same optimizations paths and not only the last `else` block, then yes let's try making it use this new templated method as well. Same goes for `Half`.

Reply:
Let's also use `static_cast<cast_scalar_t>` for the casting.

Reply:
`Half` can't benefit from the optimized path, as `sqrt` & `rqsrt` can't be enabled for `Float16` due to lack of AVX2 vectorization support for it. So, I'd only be able to add support for `BFloat16`.